File: Backend/BlockchainServices/scripts/deploy.py
"""
Automated Smart Contract Deployment Script for BuildBear
This will compile and deploy the ActivityLog contract automatically
"""
from web3 import Web3
from eth_account import Account
from solcx import compile_source, install_solc
import json
import os

# ============================================
# CONFIGURATION
# ============================================
BUILDBEAR_RPC_URL = "https://rpc.buildbear.io/nutty-darkphoenix-eda50421"
PRIVATE_KEY = "3f2eb6735d6d2ff3ee4c3db83d2f84867f7530de32b07c7ecee5e37713c536bd"

# ============================================
# SMART CONTRACT SOURCE CODE
# ============================================
CONTRACT_SOURCE = '''
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract ActivityLog {
    struct Log {
        uint256 logId;
        string serviceIdentifier;
        string action;
        string entityType;
        uint256 entityId;
        string actorUsername;
        address actorAddress;
        string changeDescription;
        string dataHash;
        uint256 timestamp;
    }

    mapping(uint256 => Log) public activityLogs;
    uint256 private logCounter;

    event ActivityLogged(
        uint256 indexed logId,
        string serviceIdentifier,
        string action,
        address indexed actorAddress
    );

    function logActivity(
        string memory _serviceIdentifier,
        string memory _action,
        string memory _entityType,
        uint256 _entityId,
        string memory _actorUsername,
        string memory _changeDescription,
        string memory _dataHash
    ) public returns (uint256) {
        uint256 newLogId = logCounter;
        
        activityLogs[newLogId] = Log({
            logId: newLogId,
            serviceIdentifier: _serviceIdentifier,
            action: _action,
            entityType: _entityType,
            entityId: _entityId,
            actorUsername: _actorUsername,
            actorAddress: msg.sender,
            changeDescription: _changeDescription,
            dataHash: _dataHash,
            timestamp: block.timestamp
        });

        emit ActivityLogged(newLogId, _serviceIdentifier, _action, msg.sender);
        
        logCounter++;
        return newLogId;
    }

    function getLogCount() public view returns (uint256) {
        return logCounter;
    }
}
'''

# ============================================
# DEPLOYMENT SCRIPT
# ============================================
def deploy_contract():
    print("=" * 60)
    print("🚀 DEPLOYING ACTIVITYLOG CONTRACT TO BUILDBEAR")
    print("=" * 60)
    
    # Step 1: Connect to BuildBear
    print("\n📡 Step 1: Connecting to BuildBear...")
    w3 = Web3(Web3.HTTPProvider(BUILDBEAR_RPC_URL))
    
    if not w3.is_connected():
        print("❌ Failed to connect to BuildBear!")
        return
    
    account = Account.from_key(PRIVATE_KEY)
    balance = w3.eth.get_balance(account.address)
    
    print(f"✅ Connected to BuildBear")
    print(f"   Account: {account.address}")
    print(f"   Balance: {w3.from_wei(balance, 'ether')} ETH")
    
    if balance == 0:
        print("⚠️  Warning: Account has 0 balance. You may need to fund it.")
    
    # Step 2: Install and compile Solidity
    print("\n🔧 Step 2: Installing Solidity compiler...")
    try:
        # Try to install 0.8.0, if it fails, try other versions
        try:
            install_solc('0.8.0')
            print("✅ Solidity compiler 0.8.0 installed")
        except Exception as e:
            print(f"⚠️  0.8.0 installation failed, trying 0.8.19...")
            try:
                install_solc('0.8.19')
                print("✅ Solidity compiler 0.8.19 installed")
            except Exception as e2:
                print(f"⚠️  0.8.19 installation failed, trying latest available...")
                # Get available versions and install the latest
                available_versions = install_solc.get_installable_solc_versions()
                if available_versions:
                    latest_version = max(available_versions)
                    install_solc(latest_version)
                    print(f"✅ Solidity compiler {latest_version} installed")
                else:
                    raise Exception("No Solidity versions available for installation")
    except Exception as e:
        print(f"❌ Failed to install Solidity compiler: {e}")
        return
    
    print("\n⚙️  Step 3: Compiling smart contract...")
    try:
        # Try compiling with the installed version
        compiled_sol = compile_source(CONTRACT_SOURCE, output_values=['abi', 'bin'])
        contract_id, contract_interface = compiled_sol.popitem()
        bytecode = contract_interface['bin']
        abi = contract_interface['abi']
        print("✅ Contract compiled successfully")
    except Exception as e:
        print(f"❌ Compilation failed: {e}")
        print("ℹ️  Trying to install and use a different Solidity version...")
        try:
            # Try installing a different version and compile again
            install_solc('0.8.19')
            compiled_sol = compile_source(CONTRACT_SOURCE, output_values=['abi', 'bin'])
            contract_id, contract_interface = compiled_sol.popitem()
            bytecode = contract_interface['bin']
            abi = contract_interface['abi']
            print("✅ Contract compiled successfully with 0.8.19")
        except Exception as e2:
            print(f"❌ Compilation still failed: {e2}")
            return
    
    # Step 4: Deploy contract
    print("\n📤 Step 4: Deploying contract to blockchain...")
    try:
        Contract = w3.eth.contract(abi=abi, bytecode=bytecode)
        
        # Get nonce
        nonce = w3.eth.get_transaction_count(account.address)
        
        # Build transaction
        transaction = Contract.constructor().build_transaction({
            'from': account.address,
            'nonce': nonce,
            'gas': 3000000,
            'gasPrice': w3.eth.gas_price
        })
        
        # Sign transaction
        signed_txn = account.sign_transaction(transaction)

        # Send transaction
        print("   Sending transaction...")
        tx_hash = w3.eth.send_raw_transaction(signed_txn.raw_transaction)
        print(f"   Transaction hash: {tx_hash.hex()}")
        
        # Wait for receipt
        print("   Waiting for confirmation...")
        tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=120)
        
        contract_address = tx_receipt.contractAddress
        
        print("\n" + "=" * 60)
        print("🎉 CONTRACT DEPLOYED SUCCESSFULLY!")
        print("=" * 60)
        print(f"\n📍 Contract Address: {contract_address}")
        print(f"🔗 Transaction Hash: {tx_hash.hex()}")
        print(f"📦 Block Number: {tx_receipt.blockNumber}")
        print(f"⛽ Gas Used: {tx_receipt.gasUsed}")
        
        # Step 5: Save to .env file
        print("\n💾 Step 5: Saving to .env file...")
        env_content = f"""
# BuildBear Blockchain Configuration
BUILDBEAR_RPC_URL={BUILDBEAR_RPC_URL}
PRIVATE_KEY={PRIVATE_KEY}
CONTRACT_ADDRESS={contract_address}
"""
        
        with open('.env', 'w') as f:
            f.write(env_content.strip())
        
        print("✅ Configuration saved to .env file")
        
        # Step 6: Save ABI
        print("\n💾 Step 6: Saving contract ABI...")
        with open('contract_abi.json', 'w') as f:
            json.dump(abi, f, indent=2)
        print("✅ ABI saved to contract_abi.json")
        
        print("\n" + "=" * 60)
        print("✨ DEPLOYMENT COMPLETE!")
        print("=" * 60)
        print("\n📋 Next Steps:")
        print("1. Set environment variable:")
        print(f"   $env:CONTRACT_ADDRESS = '{contract_address}'")
        print("\n2. Or restart your service to load from .env file")
        print("\n3. Test the blockchain service:")
        print("   python main.py")
        
        return contract_address
        
    except Exception as e:
        print(f"❌ Deployment failed: {e}")
        import traceback
        traceback.print_exc()
        return None

# ============================================
# MAIN EXECUTION
# ============================================
if __name__ == "__main__":
    print("\n⚠️  IMPORTANT: Make sure you have installed required packages:")
    print("   pip install web3 py-solc-x eth-account")
    print("\nPress Enter to continue or Ctrl+C to cancel...")
    input()
    
    contract_address = deploy_contract()
    
    if contract_address:
        print("\n✅ You can now use your blockchain service!")
    else:
        print("\n❌ Deployment failed. Please check the errors above.")

File: Backend/BlockchainServices/scripts/deploy.js
const hre = require("hardhat");

async function main() {
  console.log("\n" + "=".repeat(70));
  console.log("🚀 Deploying ActivityLogger contract to BuildBear...");
  console.log("=".repeat(70) + "\n");

  // Get the signer (your account)
  const [deployer] = await hre.ethers.getSigners();
  
  console.log("📍 Deploying from account:", deployer.address);
  const balance = await hre.ethers.provider.getBalance(deployer.address);
  console.log("💰 Account balance:", hre.ethers.formatEther(balance), "ETH\n");
  
  if (balance === 0n) {
    console.log("⚠️  WARNING: Account has 0 balance!");
    console.log("   Please fund your account on BuildBear Faucet\n");
  }

  // Get the contract factory
  console.log("📝 Getting contract factory...");
  const ActivityLogger = await hre.ethers.getContractFactory("ActivityLogger");
  
  // Deploy the contract
  console.log("🚀 Deploying contract...");
  const activityLogger = await ActivityLogger.deploy();
  
  // Wait for deployment to complete
  console.log("⏳ Waiting for deployment confirmation...");
  await activityLogger.waitForDeployment();
  
  const contractAddress = await activityLogger.getAddress();
  const deploymentTx = activityLogger.deploymentTransaction();
  
  console.log("\n" + "=".repeat(70));
  console.log("✅ CONTRACT DEPLOYED SUCCESSFULLY!");
  console.log("=".repeat(70));
  console.log(`\n📍 Contract Address: ${contractAddress}`);
  console.log(`🔗 Transaction Hash: ${deploymentTx.hash}`);
  console.log(`⛽ Gas Used: ${deploymentTx.gasLimit.toString()}`);
  console.log("=".repeat(70));
  
  // Test the contract
  console.log("\n🧪 Testing contract functionality...");
  try {
    // Test 1: Get initial log count
    const initialCount = await activityLogger.logCount();
    console.log(`✅ Initial log count: ${initialCount}`);
    
    // Test 2: Create a test log
    console.log("\n📝 Creating test log...");
    const testTx = await activityLogger.logActivity(
      "DEPLOYMENT_TEST",
      "CREATE",
      "Contract",
      1,
      "system",
      "Contract deployed and tested successfully",
      "0x" + "0".repeat(64) // Dummy hash for testing
    );
    
    console.log(`⏳ Waiting for transaction confirmation...`);
    const receipt = await testTx.wait();
    console.log(`✅ Test log created! Transaction: ${receipt.hash}`);
    
    // Test 3: Verify log was created
    const newCount = await activityLogger.logCount();
    console.log(`✅ New log count: ${newCount}`);
    
    // Test 4: Retrieve the log
    const log = await activityLogger.getLog(0);
    console.log(`✅ Log retrieved successfully`);
    console.log(`   Service: ${log.serviceIdentifier}`);
    console.log(`   Action: ${log.action}`);
    console.log(`   Actor: ${log.actorUsername}`);
    
    console.log("\n🎉 All tests passed!");
    
  } catch (error) {
    console.log("\n⚠️  Test failed:", error.message);
  }
  
  // Save configuration
  console.log("\n" + "=".repeat(70));
  console.log("📋 CONFIGURATION INSTRUCTIONS");
  console.log("=".repeat(70));
  console.log("\n1️⃣  Update your .env file with:");
  console.log(`   CONTRACT_ADDRESS=${contractAddress}`);
  console.log("\n2️⃣  Restart your FastAPI blockchain service (port 9005)");
  console.log("\n3️⃣  Test the API:");
  console.log("   • Status: http://localhost:9005/blockchain/status");
  console.log("   • API Docs: http://localhost:9005/docs");
  console.log("\n4️⃣  View on BuildBear Explorer:"); 
  console.log(`   https://rpc.buildbear.io/deaf-warlock-ac333142/address/${contractAddress}`);
  
  console.log("\n" + "=".repeat(70));
  console.log("✨ DEPLOYMENT COMPLETE!");
  console.log("=".repeat(70) + "\n");
  
  return contractAddress;
}

// Execute deployment
main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error("\n❌ DEPLOYMENT FAILED:");
    console.error(error);
    process.exit(1);
  });